src/sinks/file/mod.rs

```diff
@@ -88,98 +98,125 @@ impl PartitionedFileSink {
+                    debug!("Receiver exausted, terminating the processing loop");
+                    break;
                 }
-            };
-
-            all_files_ready = all_files_ready || ready;
-        }
-
-        for (key, last_accessed) in &self.last_accessed {
-            if last_accessed.elapsed().as_secs() > self.idle_timeout_secs {
-                if let Some(file) = self.partitions.remove(key) {
-                    let file_path = String::from_utf8_lossy(&key[..]);
-                    debug!(message = "Closing file.", file = %file_path);
-                    self.closing.insert(key.clone(), file);
+                Either::Left((Some(event), _)) => {
+                    let path = match self.partition_event(&event) {
+                        Some(path) => path,
+                        None => {
+                            // We weren't able to find the path to use for the
+                            // file.
+                            // This is already logged at `partition_event`, so
+                            // here we just skip the event.
+                            warn!("Unable to partition an event, dropping it");
+                            continue;
+                        }
+                    };
+
+                    let next_deadline = self.deadline_at();
+                    trace!(next_deadline = ?next_deadline, "Computed next deadline");
+
+                    let file = if let Some(file) = self.files.reset_at(&path, next_deadline) {
+                        trace!(path = ?path, "Working with an already opened file");
+                        file
+                    } else {
+                        trace!(path = ?path, "Opening new file");
+                        let file = match open_file(BytesPath::new(path.clone())).await {
+                            Ok(file) => file,
+                            Err(err) => {
+                                // We coundn't open the file for this event.
+                                // Maybe other events will work though! Just log
+                                // the error and skip this event.
+                                error!(path = ?path, "Unable to open the file: {}", err);
+                                continue;
+                            }
+                        };
+                        self.files.insert_at(path.clone(), (file, next_deadline));
+                        self.files.get_mut(&path).unwrap()
+                    };
+
+                    trace!(path = ?path, "Writing an event to file");
+                    if let Err(err) = write_event_to_file(file, event, &self.encoding).await {
+                        error!(path = ?path, "Failed to write file: {}", err);
+                    }
+                }
+                Either::Right((None, _)) => {
+                    // Expiration queue is empty, whatever.
+                    debug!("File expiration queue empty");
```

Comment:
this would mean no events have been fed through prob make this a rate limited trace?

Reply:
I thought this has different semantics, and it looks like that's the case from what I see in the code. I'll make sure to add tests for it at `ExpiringHashMap`.

Reply:
You are right! Tests show that it does indeed return `None` if the queue is empty. I think it is not a good fit for our use case.

If would make more sense to make it so `ExpiringHashMap::poll_expired` returns just `Poll<Result<ExpiredItem<K, V>, Error>>`, and for it to be pending if the map - and inner queue - are empty.
This will require us to add a custom waker, We'll return it if the queue returns `None`, and wake it when an element is inserted at an empty queue. It's a bit of overhead, but at least it's simple and better than making the consumer spinning on `None` - as what would be the case currently.

So, what do you think?

Reply:
That may work, can we use something like https://docs.rs/tokio/0.2.13/tokio/sync/struct.Notify.html?

Reply:
I've already implemented this, check the last few commits! I used `futures::task::AtomicWaker`. Looks like it's a very similar thing, but I like the API a bit better - feel like it's more suited for our use case.

Reply:
Which one is? I don't see why we need to use a low level type like an atomic waker? In reality we probably shouldn't even expose a pollfn!

Reply:
On a second, since we don't `Pin` when we poll, maybe it's better to use `Notify`... I'll add some tests to see if we'll have issues with the current implementation, or if we have to switch to `Notfiy`. The problematic scenario is when the future if polled from multiple call-sites - but I doubt it can realistically happen with the mutability rules we impose. I can't think of a way to accurately prove the possibility or impossibility of it atm, but I'll spend some time thinking about it. Anyhow, my opinion is that we should provide as good test coverage as possible here.

Btw, I don't think atomic waker is more low-level than notify. Fundamentally they solve the same issues but adopting `Notify` for the pollfn might be too much work.
In that light, getting rid of raw pollfn and switching to exposing a future *may be* justified. However, estimating the amount of code we'll need for that, and anticipated changes in the API - I doubt it'll be worth it. The benefit is that we would move the whole implementation into a relatively simpler (due to stricter interfaces) domain of futures. It will introduce some runtime overhead, but, ultimately, I think we can ignore that if the interface is significantly better. But, well, that's the problem - I don't think we can make it better.

Or did you mean that we should just hide the pollfn from the interface and only leave stream impl? That would somewhat improve the API. But based on the fact that `DelayQueue` itself exposes pollfn - I don't think we should try to do something fancier, especially if we can. That is, if we can - but I'll add some more tests to check edge cases.

Reply:
We should basically try to avoid as much poll based fn as possible. This is what notify enables, it gives you a waker like api that you can just use directly in an async fn.

Reply:
I mean, we implement `Stream` here, and it works atop of a pollfn internally. Should we hide the custom pollfn but leave `Stream` impl, or just get rid of both? If so, what to add instead - a `Future` impl, or an fn that returns `Future`? I was thinking about the latter, and this is what we'll probably have to use if we want to use `Notify`.
So, essentially, the significantly different cases here are:
- to have `ExpiringHashMap` implement polling on itself somehow (via `Stream`, `Future` or custom pollfn);
- to add an fn to `ExpiringHashMap` that would return `Future` or `Stream`.

What are the differences between the two? I guess, a function would be better, cause it would be able to lock the required for polling. But then - we won't possibly have multiple wakers for sure, and therefore we won't need `Notify`, as `AsyncWaker` would suffice. Otherwise, `Notify` is useful, but we need to put the handles somewhere.

Wait. If we add an fn that returns a `Future` or `Stream`, and locks the `ExpiringHashMap` under a mut ref, we won't be able to add elements into it in the other select branches, right? I think we explicitly need the `ExpiringHashMap` to be pollable, not some derived object.

Reply:
I've implemented the API we settled on!